### ISSUE TITLE
Regression with Micronaut Data and JDK 11: "com.oracle.svm.hosted.substitute.DeletedElementException: Unsupported type jdk.internal.reflect.MethodAccessorGenerator is reachable"

### ISSUE THREAD

**Issue by ilopmar** [oracle/graal#2923]
**Describe the issue**

Our CI has detected a regression introduced by this commit c616f346ae4552c04be74fe927c81f6f45fd6e6e and some Micronaut Data JDBC/JPA apps. In particular for **Oracle**, **MySql**, **MariaDB** and **Postgres**. It works for H2 and SQL Server.

This only happens with **JDK11** and it works for all the databases with JDK8.

For reference, this is the run that failed: https://gitlab.com/micronaut-projects/micronaut-graal-tests/-/pipelines/203027019

**Steps to reproduce the issue**

- `git clone https://github.com/micronaut-graal-tests/micronaut-data-jdbc-graal`
- `cd micronaut-data-jdbc-graal`
- `git checkout graalvm-issue-jdk11`
- Use GraalVM compiled from master branch with JDK11
- `./build-native-image.sh`

**Describe GraalVM and your environment:**
 - GraalVM version: **`master` branch, commit c616f346ae4552c04be74fe927c81f6f45fd6e6e**
 - JDK major version: **Only 11, it work with 8**
 - OS: **Linux Mint 20**
 - Architecture: **AMD64**

**More details**

This is the original exception:

```
Error: com.oracle.svm.hosted.substitute.DeletedElementException: Unsupported type jdk.internal.reflect.MethodAccessorGenerator is reachable
To diagnose the issue, you can add the option --report-unsupported-elements-at-runtime. The unsupported element is then reported at run time when it is accessed the first time.
Detailed message:
Trace: 
	at parsing jdk.internal.reflect.ReflectionFactory.generateConstructor(ReflectionFactory.java:513)
Call path from entry point to jdk.internal.reflect.ReflectionFactory.generateConstructor(Class, Constructor): 
	at jdk.internal.reflect.ReflectionFactory.generateConstructor(ReflectionFactory.java:513)
	at jdk.internal.reflect.ReflectionFactory.newConstructorForSerialization(ReflectionFactory.java:506)
	at java.io.ObjectStreamClass.getSerializableConstructor(ObjectStreamClass.java:1531)
	at java.io.ObjectStreamClass$2.run(ObjectStreamClass.java:509)
	at java.io.ObjectStreamClass$2.run(ObjectStreamClass.java:484)
	at com.oracle.svm.core.jdk.Target_java_security_AccessController.doPrivileged(SecuritySubstitutions.java:84)
	at sun.security.pkcs11.SunPKCS11.initToken(SunPKCS11.java:1028)
	at sun.security.pkcs11.SunPKCS11$TokenPoller.run(SunPKCS11.java:852)
	at java.lang.Thread.run(Thread.java:834)
	at com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:519)
	at com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:192)
	at com.oracle.svm.core.code.IsolateEnterStub.PosixJavaThreads_pthreadStartRoutine_e1f4a8c0039f8337338252cd8734f63a79b5e3df(generated:0)
Error: Use -H:+ReportExceptionStackTraces to print stacktrace of underlying exception
Error: Image build request failed with exit status 1
```

Adding `-H:+ReportExceptionStackTraces` the error is:

```
[data-jdbc-postgres:92885]    classlist:   4,768.80 ms,  0.96 GB
[data-jdbc-postgres:92885]        (cap):     965.81 ms,  0.96 GB
[data-jdbc-postgres:92885]        setup:   3,366.58 ms,  0.96 GB
[data-jdbc-postgres:92885]     (clinit):   1,378.49 ms,  5.33 GB
[data-jdbc-postgres:92885]   (typeflow):  41,680.42 ms,  5.33 GB
[data-jdbc-postgres:92885]    (objects):  41,868.39 ms,  5.33 GB
[data-jdbc-postgres:92885]   (features):   3,126.64 ms,  5.33 GB
[data-jdbc-postgres:92885]     analysis:  91,744.56 ms,  5.33 GB
Error: com.oracle.svm.hosted.substitute.DeletedElementException: Unsupported type jdk.internal.reflect.MethodAccessorGenerator is reachable
To diagnose the issue, you can add the option --report-unsupported-elements-at-runtime. The unsupported element is then reported at run time when it is accessed the first time.
Detailed message:
Trace: 
	at parsing jdk.internal.reflect.ReflectionFactory.generateConstructor(ReflectionFactory.java:513)
Call path from entry point to jdk.internal.reflect.ReflectionFactory.generateConstructor(Class, Constructor): 
	at jdk.internal.reflect.ReflectionFactory.generateConstructor(ReflectionFactory.java:513)
	at jdk.internal.reflect.ReflectionFactory.newConstructorForSerialization(ReflectionFactory.java:506)
	at java.io.ObjectStreamClass.getSerializableConstructor(ObjectStreamClass.java:1531)
	at java.io.ObjectStreamClass$2.run(ObjectStreamClass.java:509)
	at java.io.ObjectStreamClass$2.run(ObjectStreamClass.java:484)
	at com.oracle.svm.core.jdk.Target_java_security_AccessController.doPrivileged(SecuritySubstitutions.java:84)
	at sun.security.pkcs11.SunPKCS11.initToken(SunPKCS11.java:1028)
	at sun.security.pkcs11.SunPKCS11$TokenPoller.run(SunPKCS11.java:852)
	at java.lang.Thread.run(Thread.java:834)
	at com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:519)
	at com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:192)
	at com.oracle.svm.core.code.IsolateEnterStub.PosixJavaThreads_pthreadStartRoutine_e1f4a8c0039f8337338252cd8734f63a79b5e3df(generated:0)

com.oracle.svm.core.util.UserError$UserException: com.oracle.svm.hosted.substitute.DeletedElementException: Unsupported type jdk.internal.reflect.MethodAccessorGenerator is reachable
To diagnose the issue, you can add the option --report-unsupported-elements-at-runtime. The unsupported element is then reported at run time when it is accessed the first time.
Detailed message:
Trace: 
	at parsing jdk.internal.reflect.ReflectionFactory.generateConstructor(ReflectionFactory.java:513)
Call path from entry point to jdk.internal.reflect.ReflectionFactory.generateConstructor(Class, Constructor): 
	at jdk.internal.reflect.ReflectionFactory.generateConstructor(ReflectionFactory.java:513)
	at jdk.internal.reflect.ReflectionFactory.newConstructorForSerialization(ReflectionFactory.java:506)
	at java.io.ObjectStreamClass.getSerializableConstructor(ObjectStreamClass.java:1531)
	at java.io.ObjectStreamClass$2.run(ObjectStreamClass.java:509)
	at java.io.ObjectStreamClass$2.run(ObjectStreamClass.java:484)
	at com.oracle.svm.core.jdk.Target_java_security_AccessController.doPrivileged(SecuritySubstitutions.java:84)
	at sun.security.pkcs11.SunPKCS11.initToken(SunPKCS11.java:1028)
	at sun.security.pkcs11.SunPKCS11$TokenPoller.run(SunPKCS11.java:852)
	at java.lang.Thread.run(Thread.java:834)
	at com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:519)
	at com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:192)
	at com.oracle.svm.core.code.IsolateEnterStub.PosixJavaThreads_pthreadStartRoutine_e1f4a8c0039f8337338252cd8734f63a79b5e3df(generated:0)

	at com.oracle.svm.core.util.UserError.abort(UserError.java:82)
	at com.oracle.svm.hosted.FallbackFeature.reportAsFallback(FallbackFeature.java:217)
	at com.oracle.svm.hosted.NativeImageGenerator.runPointsToAnalysis(NativeImageGenerator.java:770)
	at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:560)
	at com.oracle.svm.hosted.NativeImageGenerator.lambda$run$0(NativeImageGenerator.java:473)
	at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(ForkJoinTask.java:1407)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)
Caused by: com.oracle.graal.pointsto.constraints.UnsupportedFeatureException: com.oracle.svm.hosted.substitute.DeletedElementException: Unsupported type jdk.internal.reflect.MethodAccessorGenerator is reachable
To diagnose the issue, you can add the option --report-unsupported-elements-at-runtime. The unsupported element is then reported at run time when it is accessed the first time.
Detailed message:
Trace: 
	at parsing jdk.internal.reflect.ReflectionFactory.generateConstructor(ReflectionFactory.java:513)
Call path from entry point to jdk.internal.reflect.ReflectionFactory.generateConstructor(Class, Constructor): 
	at jdk.internal.reflect.ReflectionFactory.generateConstructor(ReflectionFactory.java:513)
	at jdk.internal.reflect.ReflectionFactory.newConstructorForSerialization(ReflectionFactory.java:506)
	at java.io.ObjectStreamClass.getSerializableConstructor(ObjectStreamClass.java:1531)
	at java.io.ObjectStreamClass$2.run(ObjectStreamClass.java:509)
	at java.io.ObjectStreamClass$2.run(ObjectStreamClass.java:484)
	at com.oracle.svm.core.jdk.Target_java_security_AccessController.doPrivileged(SecuritySubstitutions.java:84)
	at sun.security.pkcs11.SunPKCS11.initToken(SunPKCS11.java:1028)
	at sun.security.pkcs11.SunPKCS11$TokenPoller.run(SunPKCS11.java:852)
	at java.lang.Thread.run(Thread.java:834)
	at com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:519)
	at com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:192)
	at com.oracle.svm.core.code.IsolateEnterStub.PosixJavaThreads_pthreadStartRoutine_e1f4a8c0039f8337338252cd8734f63a79b5e3df(generated:0)

	at com.oracle.graal.pointsto.constraints.UnsupportedFeatures.report(UnsupportedFeatures.java:126)
	at com.oracle.svm.hosted.NativeImageGenerator.runPointsToAnalysis(NativeImageGenerator.java:767)
	... 8 more
Caused by: com.oracle.svm.hosted.substitute.DeletedElementException: Unsupported type jdk.internal.reflect.MethodAccessorGenerator is reachable
To diagnose the issue, you can add the option --report-unsupported-elements-at-runtime. The unsupported element is then reported at run time when it is accessed the first time.
	at com.oracle.svm.hosted.substitute.AnnotationSubstitutionProcessor.lookup(AnnotationSubstitutionProcessor.java:118)
	at com.oracle.graal.pointsto.infrastructure.SubstitutionProcessor$ChainedSubstitutionProcessor.lookup(SubstitutionProcessor.java:113)
	at com.oracle.graal.pointsto.infrastructure.SubstitutionProcessor$ChainedSubstitutionProcessor.lookup(SubstitutionProcessor.java:113)
	at com.oracle.graal.pointsto.meta.AnalysisUniverse.lookupAllowUnresolved(AnalysisUniverse.java:207)
	at com.oracle.graal.pointsto.infrastructure.WrappedConstantPool.lookupType(WrappedConstantPool.java:164)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.lookupType(BytecodeParser.java:4261)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.genNewInstance(BytecodeParser.java:4578)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.processBytecode(BytecodeParser.java:5376)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.iterateBytecodesForBlock(BytecodeParser.java:3410)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.processBlock(BytecodeParser.java:3217)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.build(BytecodeParser.java:1093)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.buildRootMethod(BytecodeParser.java:987)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.GraphBuilderPhase$Instance.run(GraphBuilderPhase.java:84)
	at com.oracle.svm.hosted.phases.SharedGraphBuilderPhase.run(SharedGraphBuilderPhase.java:70)
	at jdk.internal.vm.compiler/org.graalvm.compiler.phases.Phase.run(Phase.java:49)
	at jdk.internal.vm.compiler/org.graalvm.compiler.phases.BasePhase.apply(BasePhase.java:212)
	at jdk.internal.vm.compiler/org.graalvm.compiler.phases.Phase.apply(Phase.java:42)
	at jdk.internal.vm.compiler/org.graalvm.compiler.phases.Phase.apply(Phase.java:38)
	at com.oracle.graal.pointsto.flow.MethodTypeFlowBuilder.parse(MethodTypeFlowBuilder.java:223)
	at com.oracle.graal.pointsto.flow.MethodTypeFlowBuilder.apply(MethodTypeFlowBuilder.java:357)
	at com.oracle.graal.pointsto.flow.MethodTypeFlow.doParse(MethodTypeFlow.java:313)
	at com.oracle.graal.pointsto.flow.MethodTypeFlow.ensureParsed(MethodTypeFlow.java:302)
	at com.oracle.graal.pointsto.flow.MethodTypeFlow.addContext(MethodTypeFlow.java:103)
	at com.oracle.graal.pointsto.DefaultAnalysisPolicy$DefaultSpecialInvokeTypeFlow.onObservedUpdate(DefaultAnalysisPolicy.java:368)
	at com.oracle.graal.pointsto.flow.TypeFlow.notifyObservers(TypeFlow.java:470)
	at com.oracle.graal.pointsto.flow.TypeFlow.update(TypeFlow.java:542)
	at com.oracle.graal.pointsto.BigBang$2.run(BigBang.java:547)
	at com.oracle.graal.pointsto.util.CompletionExecutor.lambda$execute$0(CompletionExecutor.java:173)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1426)
	... 5 more
Error: Image build request failed with exit status 1
```

Adding `--report-unsupported-elements-at-runtime` makes it work again but I don't see this as a solution but as a temporary workaround because this flag wasn't necessary before for the mentioned databases. 
It still works for H2 and Sql Server because for these two, we are using the flag.

**Comment by ilopmar** [comment]
Now that the change has been "propagated" to the daily dev-releases, our jobs for those releases started to fail too: https://gitlab.com/micronaut-projects/micronaut-graal-tests/-/pipelines/204324955

**Comment by d-kozak** [comment]
Hello. 
In the PR that caused the regression, I've added support for lookup of services registered using module descriptors. I've debugged what is happening in your case and the reason is the following. The service  **javax.naming.spi.InitialContextFactory** is reachable, which makes it's two providers **com.sun.jndi.rmi.registry.RegistryContextFactory** and **com.sun.jndi.dns.DnsContextFactory** reachable as well.

Out of those, the **com.sun.jndi.rmi.registry.RegistryContextFactory** causes the **sun.security.pkcs11.SunPKCS11** to be reachable by some transitive dependency. Seems that the pkcs class is invalid because it uses object serialization. Once the PR for that is merged, the problem might disappear. Until that, we can fix it by excluding either the service or its rmi provider. 

My question is, are you using any of the mentioned classes in your project? Just because it compiles without it does not mean all is well, so I just wanted to discuss that with you.

**Comment by ilopmar** [comment]
Hi @d-kozak,

I'm not using any of those classes directly, no idea why they are used, maybe because of some dependency. We have some initializations for some `com.sun` classes that may trigger those other classes, but TBH, I'm not sure. We've been using those for a long time:

- https://github.com/micronaut-projects/micronaut-core/blob/b5cc40ea85d020d875f5a95e570188d4a47890a1/http-netty/src/main/resources/META-INF/native-image/io.micronaut/netty/native-image.properties
- https://github.com/micronaut-projects/micronaut-core/blob/9b2db286b460febe526f330528e5a09faa24058d/runtime/src/main/resources/META-INF/native-image/io.micronaut/runtime-graal/native-image.properties

I didn't follow all you said. Is there any PR submitted I can test locally? 
How can I exclude the service or the rmi provider?

**Comment by d-kozak** [comment]
The PR I mentioned is this one https://github.com/oracle/graal/pull/2794 - Its aim is to add support for object serialization to the native image. According to the stack trace, it seems that the problem with **sun.security.pkcs11.SunPKCS11** is related to object serialization. So it might be fixed when it is merged.

Regarding the other question, these options are currently not part of the API. I'll discuss it with others today whether we should add it. 

In any case, I apologize for any inconvenience caused by this problem. I am working on it and I'll try to fix it as soon as possible.

**Comment by ilopmar** [comment]
I've tried compiling from source code using that PR and it works now. I see that is still draft so if you want me to give it another try once is ready and before merging it, please let me know.

> In any case, I apologize for any inconvenience caused by this problem. I am working on it and I'll try to fix it as soon as possible.

No need to apologize. There is time to fix this before releasing 20.3 :smiley:

**Comment by ilopmar** [comment]
I see that there is already a branch for the 20.3 release. I don't want to be annoying, just wanted to ask if you folks think this could be fixed and included in that release.

**Comment by fernando-valdez** [comment]
Hi @ilopmar, I will test it today using the latest 20.3 build. 
I'll be back to you once I have any news

**Comment by ilopmar** [comment]
@fernando-valdez you mean that the fix is already included in the `release/graal-vm/20.3` branch?

**Comment by fernando-valdez** [comment]
no, I don't know if the fix was included in the latest build. But there are a lot of fixes, and I want to see if your issue was solved by any chance.

**Comment by ilopmar** [comment]
Ahhh, I see...

**Comment by d-kozak** [comment]
Hello @ilopmar,

I've merged a fix for this issue recently. I checked your repo again and the image build was successful. Please let me know if it works in your environment as well.

**Comment by ilopmar** [comment]
(@d-kozak I just writing this when you put your comment)

I can confirm that the latest changes in 20.3 release branch fixed this issue. The CI is (almost) back to green: https://gitlab.com/micronaut-projects/micronaut-graal-tests/-/pipelines/212657011

P.S: The only issue left is #2940, it would be great if that can also be fixed before 20.3
